### ISSUE TITLE
Use ReverseDiff for J*v calculations when factorization is a Krylov method?

### ISSUE THREAD

**Issue by ChrisRackauckas** [JuliaDiffEq/OrdinaryDiffEq.jl#396]
out of curiosity, are there special tricks that can be done to compute `(I - gamma*J)*v` without forming `J` (where `gamma` is some constant)?
Is that some kind of thing that can be done with reverse differentiation, J being the normal Jacobian (`f(u,p,t), J = df/du`)

jekbradbury [7:18 PM]
isn't `(I-gamma*J)*v` just `I*v - gamma*(J*v) = v - gamma*(J*v)` and the latter term doesn't require forming J? (edited)

Chris Rackauckas [7:19 PM]
I don't know reverse diff so this is actually new to me that it does matvecs without forming the mat
so yes, if it does J*v without the J then that's all that's needed

jekbradbury [7:19 PM]
so does forward mode
that does v*J :wink:

Chris Rackauckas [7:20 PM]
forward mode computes the `J`?
At least ForwardDiff returns `J` and then we do `J*v`?

jekbradbury [7:20 PM]
you _can_ compute J (with either forward or reverse mode)
but if you're interested in `v*J` for some `v` you can provide `v` as the seed for the vector of duals used in forward mode evaluation

Chris Rackauckas [7:21 PM]
but we can't do the other way around?

jekbradbury [7:21 PM]
and the vector of duals that comes out will be `v*J`
right
in reverse mode, you provide `v` as the seed for the vector of output sensitivities, and backpropagate it through the computation (edited)
then the vector of sensitivities that comes out on the input side is `J*v`

Chris Rackauckas [7:22 PM]
without requiring memory for the J?

jekbradbury [7:22 PM]
yeah

Chris Rackauckas [7:22 PM]
ReverseDiff.jl can do this?

jekbradbury [7:23 PM]
yes, and so can Flux, and PyTorch, etc.
the more :mind_blown: bit is the fact that you can also do this for Hessian-vector products, traditionally using one forward mode pass and one reverse mode pass but it's also possible with two fwd-mode or two reverse-mode passes

Chris Rackauckas [7:24 PM]
http://www.juliadiff.org/ReverseDiff.jl/api/
I don't see the seeding thing here?

jekbradbury [7:26 PM]
ReverseDiff itself doesn't expose an API for vector-valued functions, so the way you have to do it there is to dot the output with `v` inside your function
but `Flux.back` takes a `Tracked` and a sensitivity/vector of sensitivities

Chris Rackauckas [7:28 PM]
The flux tracked stuff is in Flux.jl proper though?
I'd be weary on taking on that whole stack as a dependency just for this
I'm onto a different topic BTW
I am now thinking of how to use this to build a PDE solver with autodiff'd `J*v` for lowered memory usage. (edited)
Is the `J*v` expensive to compute though compared to having cached `J` and then doing `J*v`? This would have to be repeated with a few `v`'s.
@yingbo_ma

Yingbo Ma [7:35 PM]
joined #machine-learning by invitation from Chris Rackauckas.

Chris Rackauckas [7:35 PM]
inviting you since I think you should read this last little bit.

jekbradbury [7:36 PM]
reverse mode autodiff is definitely the “right” way to compute `J*v` in terms of memory, flops, etc
probably even relative to caching J since I think it can take less compute than simply multiplying a precomputed J by a preexisting v?

Chris Rackauckas [7:37 PM]
interesting
so is this used in Newton methods at all?
I haven't seen it before, but for Newton-Krylov this sounds awesome.

jekbradbury [7:41 PM]
Hessian-free optimization methods including those based on the Gauss-Newton matrix use this (edited)
but again I don’t know anything about diffeqs

Chris Rackauckas [7:41 PM]
well DiffEq solvers are still using numerical diff for most Jacobians
since they were written long before widespread AD (or in most cases, even before C++) was a thing.
I just set things up to use ForwardDiff by default since they are Jacobians, many times small, so from Jarrett's talks about the diffs I didn't think ReverseDiff makes sense
and in most cases you need the Jacobian since factoring `J` speeds up the algorithm a lot
but when `J` is sufficiently large you use Krylov methods and only need `J*v`
and now it sounds like we have some magic to really help this case out.
